### ISSUE TITLE
Export to git fails with `ResourceNotFound` error

### ISSUE THREAD

**Issue by pdpinch** [mitodl/edx-platform#104]
### Steps to reproduce ##

1. Export the course at https://studio.mitx.mit.edu/export/course-v1:MITx+2.854r_1+2018_Fall
2. Import the course into studio running in devstack
3. Edit the giturl to one that you can push to
4. In Studio, "Export to Git" from the Tools menu

### Expected behavior

- Course exports to the local disk and then pushes to remote git repo

### Actual behavior

- Export fails with the error `ResourceNotFound: resource 'course/static/52a76ce9-85ed-4b1d-b853-3142662b1fdb-en.srt' not found`

**Comment by pdpinch** [comment]
It's odd that this fails when using "EXport to Git" but not with regular "Export"

**Comment by pdpinch** [comment]
We are currently observing this in production with the summer 2018 release (mitx/hawthorn branch). I haven't tested it yet with the spring 2019 release  mitx/hawthorn-release branch).

**Comment by pdpinch** [comment]
@amir-qayyum-khan can you take a look into this?

**Comment by amir-qayyum-khan** [comment]
@mitodl/devops is `/edx/var/edxapp/export_course_repos` folder available. And what is permission on it?

**Comment by amir-qayyum-khan** [comment]
@pdpinch i cannot access that course. https://studio.mitx.mit.edu/export/course-v1:MITx+2.854r_1+2018_Fall. is other courses working fine? if yes then i need to check course xml files may be there is a wrong entry somewhere.

**Comment by blarghmatey** [comment]
@amir-qayyum-khan that directory is present with the appropriate permissions. (`www-data`)

**Comment by amir-qayyum-khan** [comment]
reproduced exception

```
2018-11-13 09:27:40,542 ERROR 24763 [django.request] exception.py:135 - Internal Server Error: /export_git/course-v1:MIT+MIT01+2016
Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/exception.py", line 41, in inner
    response = get_response(request)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 249, in _legacy_get_response
    response = self._get_response(request)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 187, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 185, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 185, in inner
    return func(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 149, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/contrib/auth/decorators.py", line 23, in _wrapped_view
    return view_func(request, *args, **kwargs)
  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/export_git.py", line 44, in export_git
    request.user,
  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/git_export_utils.py", line 138, in export_to_git
    root_dir, course_dir)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/modulestore/xml_exporter.py", line 346, in export_course_to_xml
    CourseExportManager(modulestore, contentstore, course_key, root_dir, course_dir).export()
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/modulestore/xml_exporter.py", line 169, in export
    courselike.add_xml_to_node(root)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/xml_module.py", line 589, in add_xml_to_node
    super(XmlParserMixin, self).add_xml_to_node(node)  # pylint: disable=bad-super-call
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1107, in add_xml_to_node
    xml_string = self.export_to_xml(self.runtime.export_fs)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/xml_module.py", line 578, in export_to_xml
    super(XmlDescriptor, self).add_xml_to_node(node)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/xml_module.py", line 444, in add_xml_to_node
    xml_object = self.definition_to_xml(self.runtime.export_fs)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/course_module.py", line 1055, in definition_to_xml
    xml_object = super(CourseDescriptor, self).definition_to_xml(resource_fs)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/seq_module.py", line 666, in definition_to_xml
    self.runtime.add_block_as_child_node(child, xml_object)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1513, in add_block_as_child_node
    block.add_xml_to_node(child)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/xml_module.py", line 589, in add_xml_to_node
    super(XmlParserMixin, self).add_xml_to_node(node)  # pylint: disable=bad-super-call
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1107, in add_xml_to_node
    xml_string = self.export_to_xml(self.runtime.export_fs)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/xml_module.py", line 578, in export_to_xml
    super(XmlDescriptor, self).add_xml_to_node(node)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/xml_module.py", line 444, in add_xml_to_node
    xml_object = self.definition_to_xml(self.runtime.export_fs)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/seq_module.py", line 666, in definition_to_xml
    self.runtime.add_block_as_child_node(child, xml_object)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1513, in add_block_as_child_node
    block.add_xml_to_node(child)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/xml_module.py", line 589, in add_xml_to_node
    super(XmlParserMixin, self).add_xml_to_node(node)  # pylint: disable=bad-super-call
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1107, in add_xml_to_node
    xml_string = self.export_to_xml(self.runtime.export_fs)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/xml_module.py", line 578, in export_to_xml
    super(XmlDescriptor, self).add_xml_to_node(node)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/xml_module.py", line 444, in add_xml_to_node
    xml_object = self.definition_to_xml(self.runtime.export_fs)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/seq_module.py", line 666, in definition_to_xml
    self.runtime.add_block_as_child_node(child, xml_object)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1513, in add_block_as_child_node
    block.add_xml_to_node(child)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/xml_module.py", line 444, in add_xml_to_node
    xml_object = self.definition_to_xml(self.runtime.export_fs)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/vertical_block.py", line 194, in definition_to_xml
    self.runtime.add_block_as_child_node(child, xml_object)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1513, in add_block_as_child_node
    block.add_xml_to_node(child)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/xml_module.py", line 589, in add_xml_to_node
    super(XmlParserMixin, self).add_xml_to_node(node)  # pylint: disable=bad-super-call
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1107, in add_xml_to_node
    xml_string = self.export_to_xml(self.runtime.export_fs)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/xml_module.py", line 578, in export_to_xml
    super(XmlDescriptor, self).add_xml_to_node(node)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/xml_module.py", line 444, in add_xml_to_node
    xml_object = self.definition_to_xml(self.runtime.export_fs)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/video_module/video_module.py", line 735, in definition_to_xml
    course_id=unicode(self.runtime.course_id.for_branch(None))
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/edxval/api.py", line 832, in export_to_xml
    return create_transcripts_xml(video_id, video_el, resource_fs, static_dir)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/edxval/api.py", line 887, in create_transcripts_xml
    combine(u'course', static_dir)  # File system should not start from /draft directory.
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/edxval/api.py", line 854, in create_transcript_file
    create_file_in_fs(transcript_content, transcript_name, resource_fs, static_dir)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/edxval/utils.py", line 198, in create_file_in_fs
    with file_system.open(combine(static_dir, file_name), 'wb') as f:
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/fs/osfs.py", line 386, in open
    **options
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/fs/error_tools.py", line 82, in __exit__
    traceback
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/fs/osfs.py", line 386, in open
    **options
ResourceNotFound: resource 'course/static/52a76ce9-85ed-4b1d-b853-3142662b1fdb-en.srt' not found
```

**Comment by amir-qayyum-khan** [comment]
i made a fix https://github.com/amir-qayyum-khan/edx-val/pull/1. @pdpinch 
but it is in a different repo.

**Comment by pdpinch** [comment]
Sorry, I don’t even know what this repo is for? I didn’t think we ran it residentially. 

Can you explain?

**Comment by amir-qayyum-khan** [comment]
1) There is a module video_module which returns an xml string representing video module. It use pyfilesystem. It create xml and write is on storage before exporting it to git. For writing file it uses edxval
https://github.com/mitodl/edx-platform/blob/mitx/hawthorn/common/lib/xmodule/xmodule/video_module/video_module.py#L731

2) we are getting issue on edX val. Folders it is expecting are not on disk. Thats why below line crash
https://github.com/edx/edx-val/blob/0.1.14/edxval/utils.py#L198

3) i edited repo and creating folders so that xml can be create and push to git

### Results:
`course/static/52a76ce9-85ed-4b1d-b853-3142662b1fdb-en.srt` is exported after solution
https://github.com/amir-qayyum-khan/test_edx_course/tree/master/static

**Comment by pdpinch** [comment]
There is a fix to this deployed to mitx-qa. It needs to be tested.

**Comment by pdpinch** [comment]
@amir-qayyum-khan I'm trying to test the fix on mitx-qa, but I'm getting the same error when i try to export. At least, I think it's the same error. 

<img width="1641" alt="screen shot 2018-11-29 at 1 17 06 pm" src="https://user-images.githubusercontent.com/430126/49242665-5851aa80-f3d9-11e8-834b-2d03f09904fa.png">

I'll will give you access to my test course in a few minutes.

**Comment by pdpinch** [comment]
As discussed in Slack, this is actually a different error. 

I'm not sure if it's a configuration issue with our QA stack, or a bug in the code. Amir, can you try importing and exporting my test course on your devstack to see if it works, and where it write this .srt file to?

**Comment by pdpinch** [comment]
@annagav you'll have to pick this up from Amir. He has a branch that he thought would fix the issue, but in our experience so far, it has not. 

I would recommend starting by running our next release of Residential MITx in docker and see if you can reproduce the issue with the current branch. Then try applying Amir's patch to see if it addresses the issue. If/when you get it working, please open a new PR against the next release branch.

**Comment by Ayub-Khan** [comment]
Started debugging this.

**Comment by Ayub-Khan** [comment]
@pdpinch I do not have access to [Course](https://studio.mitx.mit.edu/export/course-v1:MITx+2.854r_1+2018_Fall). I did check the issue [code](https://github.com/edx/edx-val/blob/master/edxval/utils.py#L198). It seems issue might not be as described in earlier comments. Because [pyfilesystem open code](https://github.com/PyFilesystem/pyfilesystem2/blob/897b21d3b1cca00b882894b2a140be2c30be33b8/fs/osfs.py#L375) only raises resource error in case of [those errors](https://github.com/PyFilesystem/pyfilesystem2/blob/897b21d3b1cca00b882894b2a140be2c30be33b8/fs/error_tools.py#L27:L34). But to find what the actual issue is, i will need the course access

**Comment by pdpinch** [comment]
Now I can't even export the course... but you should have access to it @Ayub-Khan

**Comment by pdpinch** [comment]
@Ayub-Khan Here is a more relevant test case, on our qa-next stack. 

https://studio-mitx-qa-next.mitx.mit.edu/course/course-v1:TestX+git101+2019_Spring

In studio, this course throws an error from [export](https://studio-mitx-qa-next.mitx.mit.edu/export/course-v1:TestX+git101+2019_Spring) and [export to git](https://studio-mitx-qa-next.mitx.mit.edu/export_git/course-v1:TestX+git101+2019_Spring?action=push)

I'll send you the course tar ball so you can test with devstack.

**Comment by Ayub-Khan** [comment]
@pdpinch Still can not access the course.

**Comment by pdpinch** [comment]
You may need to quit your browser in order to get a new session from CAS. 

Probably more useful: I sent you a link in Slack to a copy of the course on Google Drive.

**Comment by pdpinch** [comment]
@Ayub-Khan, fyi, configuring export to git requires a feature flag, `ENABLE_EXPORT_GIT` and an advanced setting `gitrul` on any course to be exported. These two in combination will add an "Export to Git" option in the Tools menu. 

That said, the course I linked above is failing to "Export" as well as "Export to Git" so the issue seems to be more widespread than was originally described.

**Comment by Ayub-Khan** [comment]
Created following pr with the fix. [Pull Request](https://github.com/mitodl/edx-platform/pull/116)
@pdpinch kindly add some one for review.

**Comment by pdpinch** [comment]
@Ayub-Khan's fix appears to have worked (according to [@NotoriousMKD in Slack](https://mitodl.slack.com/archives/G2E5J5CF5/p1545948337008400). If so, we should submit it upstream to edX, but that will take more time than we have remaining from Ayub.

**Comment by pdpinch** [comment]
Hm, [2.854 export](https://studio.mitx.mit.edu/export/course-v1:MITx+2.854r_1+2018_Fall) is still busted, both the vanilla and the "to git" variety. 

The original issue was with 6.0001, but I can't find that course any more (it was probably archived).

**Comment by Ayub-Khan** [comment]
@pdpinch I just had a look it seems this is a different kind of issue.

**Comment by pdpinch** [comment]
Despite fixing #122, we have confirm that this is still an issue. 

We might still be able to unblock the export with this PR though: mitodl/edx-val#2

**Comment by tasawernawaz** [comment]
@pdpinch Taking a look into this. I am able to export this course with simple export but seeing a sever error while exporting with GIT.

**Comment by tasawernawaz** [comment]
@pdpinch I am not able to reproduce the issue on my local machine, I have successfully exported course to git. See https://github.com/tasawernawaz/test_export_course
I have also tried it with one of my own local courses and did not find any issue. 

Could you please share any other course with which I can try?

P.S: thanks @shaidar for his help to config git/ssh on devstack.

**Comment by pdpinch** [comment]
Which branch of edX-platform are you testing with?

**Comment by tasawernawaz** [comment]
@pdpinch `mitx/hawthorn-release`

**Comment by pdpinch** [comment]
Can you test with the course mentioned in the issue description?

**Comment by tasawernawaz** [comment]
@pdpinch I have tested with that course, I can reconfirm this.

**Comment by tasawernawaz** [comment]
I followed following steps:

- Exported course from https://studio.mitx.mit.edu/export/course-v1:MITx+2.854r_1+2018_Fall
- Imported downloaded course to my local devstack
- Successfully exported and pushed to git.

![screen shot 2019-01-10 at 5 53 02 pm](https://user-images.githubusercontent.com/4245618/50969533-9fff4200-1500-11e9-9094-efc67e8dafbe.png)

**Comment by pdpinch** [comment]
Ok, new theory. It has something to do with the difference between this code, used by studio:

https://github.com/mitodl/edx-platform/blob/master/cms/djangoapps/contentstore/tasks.py#L647

and this code, used by export to git:

https://github.com/mitodl/edx-platform/blob/master/cms/djangoapps/contentstore/git_export_utils.py#L136-L137

In particular, I'm suspicious of the former's use of `course_module.url_name` for the last parameter and the latter's use of `course_dir`, derived from, I think, the git URL.

**Comment by tasawernawaz** [comment]
@pdpinch thank you, I will look into the code links.

**Comment by tasawernawaz** [comment]
@pdpinch Interestingly while checkout above code links, I was able reproduce this issue, then even I can't simply export and download. 

Given that, I have installed manually your branch https://github.com/mitodl/edx-val/pull/2 which eventually fixed the issue (raised exception and continue to export). 

Answer to (_In particular, I'm suspicious of the former's use of course_module.url_name for the last parameter and the latter's use of course_dir, derived from, I think, the git URL._)

You have tagged master version, in our version we have [git_export_utils.py#L134](https://github.com/mitodl/edx-platform/blob/mitx/hawthorn-release/cms/djangoapps/contentstore/git_export_utils.py#L134)

**Comment by pdpinch** [comment]
That seems like progress. 

Can you take over my branch and improve the exception text? I think if it’s the same as the log message that would be good.

I’d like to submit this val change upstream to edX. Do you think it will need additional tests for them to accept it?

**Comment by tasawernawaz** [comment]
@pdpinch what actually text message you want ? Where I can see log message ?

At this point I am not sure, if on edx-platform master we can reproduce this issue or not. If this really required we can create a PR and add tests if they required.

**Comment by pdpinch** [comment]
Take a look at https://github.com/mitodl/edx-val/blob/61a39bb20cb8e2c1242c21d40622a1802e19886c/edxval/api.py#L295-L298

We can talk about edX strategy later after we get this working.

**Comment by tasawernawaz** [comment]
@pdpinch Why we really need to duplicate a message, where exception is already logging error ? Can't we simply continue after logging exception? Why we need to raise it again ?

**Comment by tasawernawaz** [comment]
@pdpinch mitodl/edx-val#2  is closed now, can we close this now ?

**Comment by pdpinch** [comment]
The error in Export to Git no longer happens in studio.mitx, but it is still happening in studio-mitx-qa. 

I think the root cause has to be misplaced static (transcript) files. We mucked around so much with the configuration + manually moving files around, I'm not sure anymore which server is set up correctly. 

I'm going to close this issue and open another one that is specific to studio-mtix-qa.
